pkg/cache/inmemory.go

```diff
@@ -0,0 +1,302 @@
+}
+
+// NewInMemoryCacheWithConfig creates a new thread-safe LRU cache and ensures the total cache
+// size approximately does not exceed maxBytes.
+func NewInMemoryCacheWithConfig(name string, logger log.Logger, reg prometheus.Registerer, config InMemoryCacheConfig) (*InMemoryCache, error) {
+	if config.MaxItemSize > config.MaxSize {
+		return nil, errors.Errorf("max item size (%v) cannot be bigger than overall cache size (%v)", config.MaxItemSize, config.MaxSize)
+	}
+
+	c := &InMemoryCache{
+		logger:           logger,
+		maxSizeBytes:     uint64(config.MaxSize),
+		maxItemSizeBytes: uint64(config.MaxItemSize),
+	}
+
+	c.evicted = promauto.With(reg).NewCounter(prometheus.CounterOpts{
+		Name:        "thanos_cache_inmemory_items_evicted_total",
+		Help:        "Total number of items that were evicted from the inmemory cache.",
+		ConstLabels: prometheus.Labels{"name": name},
+	})
+
+	c.added = promauto.With(reg).NewCounter(prometheus.CounterOpts{
+		Name:        "thanos_cache_inmemory_items_added_total",
+		Help:        "Total number of items that were added to the inmemory cache.",
+		ConstLabels: prometheus.Labels{"name": name},
+	})
+
+	c.requests = promauto.With(reg).NewCounter(prometheus.CounterOpts{
+		Name:        "thanos_cache_inmemory_requests_total",
+		Help:        "Total number of requests to the inmemory cache.",
+		ConstLabels: prometheus.Labels{"name": name},
+	})
+
+	c.hitsExpired = promauto.With(reg).NewCounter(prometheus.CounterOpts{
+		Name:        "thanos_cache_inmemory_hits_on_expired_data_total",
+		Help:        "Total number of requests to the inmemory cache that were a hit but needed to be evicted due to TTL.",
+		ConstLabels: prometheus.Labels{"name": name},
+	})
+
+	c.overflow = promauto.With(reg).NewCounter(prometheus.CounterOpts{
+		Name:        "thanos_cache_inmemory_items_overflowed_total",
+		Help:        "Total number of items that could not be added to the inmemory cache due to being too big.",
+		ConstLabels: prometheus.Labels{"name": name},
+	})
+
+	c.hits = promauto.With(reg).NewCounter(prometheus.CounterOpts{
+		Name:        "thanos_cache_inmemory_hits_total",
+		Help:        "Total number of requests to the inmemory cache that were a hit.",
+		ConstLabels: prometheus.Labels{"name": name},
+	})
+
+	c.current = promauto.With(reg).NewGauge(prometheus.GaugeOpts{
+		Name:        "thanos_cache_inmemory_items",
+		Help:        "Current number of items in the inmemory cache.",
+		ConstLabels: prometheus.Labels{"name": name},
+	})
+
+	c.currentSize = promauto.With(reg).NewGauge(prometheus.GaugeOpts{
+		Name:        "thanos_cache_inmemory_items_size_bytes",
+		Help:        "Current byte size of items in the inmemory cache.",
+		ConstLabels: prometheus.Labels{"name": name},
+	})
+
+	c.totalCurrentSize = promauto.With(reg).NewGauge(prometheus.GaugeOpts{
+		Name: "thanos_cache_inmemory_total_size_bytes",
+		Help: "Current byte size of items (both value and key) in the inmemory cache.",
+	})
```

Comment:
Do we need both?

Reply:
I feel we do. Similar to https://github.com/thanos-io/thanos/blob/1fff9a7157d1f3b7d5e0ce36f999b10bb3ae7ec4/pkg/store/cache/inmemory.go#L48-L49 it not a bad idea to keep track of total size irrespective of how small the `key` is. Please let me know if you have other thoughts.